### ISSUE TITLE
FreestyleRet

### ISSUE THREAD

**Issue by Hyq0203** [LighteningDarkness/HIT-23spring-SystemOfCompiling#1]
Hello, brother. I saw that you left a message on the FreestyleRet project before. I would like to ask what parameters you should include in the gen_cluster. py file, such as origin_resume, gram_encoder_path, style_cluster_path

**Comment by LighteningDarkness** [comment]
Hello bro, this is my settings:
```
python src/dataset/gen_cluster.py --resume BLIP_ITM_s2i.pth --origin_resume model_large.pth --prompt BLIP_ITM
# python src/dataset/gen_cluster.py --resume FreeCLIP.pth --origin_resume openai --prompt DeepPrompt
# python src/dataset/gen_cluster.py --resume FreeCLIP.pth --origin_resume openai --prompt ShallowPrompt --prompt_location Bottom
# python src/dataset/gen_cluster.py --resume FreeCLIP.pth --origin_resume openai --prompt ShallowPrompt --prompt_location Shallow
```

It has been a long time since I run this repo last time so i am not sure whether it works and i remember that i modified some codes related to checkpoint loading. You can try it and i hope it helpful for you.

**Comment by LighteningDarkness** [comment]
and here is my gen_cluster.py. Good luck!



import cv2
from sklearn.cluster import KMeans
import argparse
import torch
import torch.nn as nn
import json
import sys
import numpy as np
from PIL import Image
from tqdm import tqdm
import os
from open_clip.transform import image_transform

import sys
sys.path.append(".")
from src.models import ShallowStyleRetrieval, DeepStyleRetrieval, BLIP_Retrieval, BLIP_ITM


def parse_args():
    parser = argparse.ArgumentParser(description='Parse args for generate cluster.')

    # project settings
    parser.add_argument('--resume', default='', type=str, help='load checkpoints from given path')
    parser.add_argument('--origin_resume', default='', type=str, help='load checkpoints from given path')
    parser.add_argument('--gram_encoder_path', default='vgg_normalised.pth', type=str, help='load vgg from given path')
    parser.add_argument('--style_cluster_path', default='pretrained/style_cluster.npy', type=str, help='load vgg from given path')
    parser.add_argument('--device', default='cuda:0')
    parser.add_argument('--num_workers', default=6, type=int)

    # data settings
    parser.add_argument("--type", type=str, default='style2image', help='choose train text2image or style2image.')
    parser.add_argument("--style", type=str, default='sketch', help='choose sketch, art or mosaic.')
    parser.add_argument("--style2", type=str, default='art', help='choose sketch, art or mosaic.')
    parser.add_argument("--test_dataset_path", type=str, default='DSR/')
    parser.add_argument("--test_json_path", type=str, default='DSR/test.json')
    parser.add_argument("--batch_size", type=int, default=16)

    # model settings
    parser.add_argument('--prompt', type=str, default='BLIP_Retrieval', choices=['ShallowPrompt','DeepPrompt', 'BLIP_Retrieval','BLIP_ITM'],help='ShallowPrompt or DeepPrompt or BLIP_Retrieval')
    parser.add_argument('--prompt_location', type=str, default='Bottom', help='Shallow or Bottom')
    parser.add_argument('--init', type=str, default='random', help='random, gram, or style')
    parser.add_argument('--gram_prompts', type=int, default=4)
    parser.add_argument('--gram_prompt_dim', type=int, default=1024)
    parser.add_argument('--style_prompts', type=int, default=4)
    parser.add_argument('--style_prompt_dim', type=int, default=1024)

    args = parser.parse_args()
    return args


root_path = 'DSR/'

image_mean = (0.48145466, 0.4578275, 0.40821073)
image_std = (0.26861954, 0.26130258, 0.27577711)

args = parse_args()

filelist = json.load(open('DSR/train.json', 'r'))

process = image_transform(224, False, image_mean, image_std)

embed = []


if args.prompt == 'ShallowPrompt':
    model = ShallowStyleRetrieval(args)
elif args.prompt == 'DeepPrompt':
    model = DeepStyleRetrieval(args)
elif args.prompt == 'BLIP_Retrieval':
    model = BLIP_Retrieval(args)
else:
    model = BLIP_ITM(args)

model.load_state_dict(torch.load(args.resume,map_location="cpu"))
model = model.to(args.device)

all_center = []
save_path = f"gram_cluster_{args.prompt}" if args.prompt != "ShallowPrompt" else f"gram_cluster_{args.prompt}_{args.prompt_location}"
os.makedirs(save_path,exist_ok=True)
for subdir in ['sketch','art','images','mosaic']:
    for file in tqdm(filelist):

        path = root_path + f'{subdir}/' + file['image']
        img = process(Image.open(path)).unsqueeze(0).to('cuda:0')
        feat = model._get_gram_prompt(img)  # (1, 4, 1024)
        embed.append(feat.detach().cpu())


    feature = torch.stack(embed, dim=0).squeeze(1).detach().numpy()
    feature = feature.reshape(feature.shape[0],-1)

    np.save(f'{save_path}/{subdir}.npy', feature)

    kmeans = KMeans(n_clusters=1)

    kmeans.fit(feature)

    cluster_labels = kmeans.labels_

    cluster_centers = kmeans.cluster_centers_

    cluster_tensor = np.mean(cluster_centers, axis=0)
    cluster_tensor = cluster_tensor.reshape((1, 4096))
    all_center.append(cluster_tensor)

all_save_path = f'style_cluster_{args.prompt}' if args.prompt != "ShallowPrompt" else f'style_cluster_{args.prompt}_{args.prompt_location}'
np.save(f'{all_save_path}.npy', np.concatenate(all_center))

**Comment by LighteningDarkness** [comment]
here is my gen_cluster.py and i hope it helpful for you. Good luck!

```
import cv2
from sklearn.cluster import KMeans
import argparse
import torch
import torch.nn as nn
import json
import sys
import numpy as np
from PIL import Image
from tqdm import tqdm
import os
from open_clip.transform import image_transform

import sys
sys.path.append(".")
from src.models import ShallowStyleRetrieval, DeepStyleRetrieval, BLIP_Retrieval, BLIP_ITM


def parse_args():
    parser = argparse.ArgumentParser(description='Parse args for generate cluster.')

    # project settings
    parser.add_argument('--resume', default='', type=str, help='load checkpoints from given path')
    parser.add_argument('--origin_resume', default='', type=str, help='load checkpoints from given path')
    parser.add_argument('--gram_encoder_path', default='vgg_normalised.pth', type=str, help='load vgg from given path')
    parser.add_argument('--style_cluster_path', default='pretrained/style_cluster.npy', type=str, help='load vgg from given path')
    parser.add_argument('--device', default='cuda:0')
    parser.add_argument('--num_workers', default=6, type=int)

    # data settings
    parser.add_argument("--type", type=str, default='style2image', help='choose train text2image or style2image.')
    parser.add_argument("--style", type=str, default='sketch', help='choose sketch, art or mosaic.')
    parser.add_argument("--style2", type=str, default='art', help='choose sketch, art or mosaic.')
    parser.add_argument("--test_dataset_path", type=str, default='DSR/')
    parser.add_argument("--test_json_path", type=str, default='DSR/test.json')
    parser.add_argument("--batch_size", type=int, default=16)

    # model settings
    parser.add_argument('--prompt', type=str, default='BLIP_Retrieval', choices=['ShallowPrompt','DeepPrompt', 'BLIP_Retrieval','BLIP_ITM'],help='ShallowPrompt or DeepPrompt or BLIP_Retrieval')
    parser.add_argument('--prompt_location', type=str, default='Bottom', help='Shallow or Bottom')
    parser.add_argument('--init', type=str, default='random', help='random, gram, or style')
    parser.add_argument('--gram_prompts', type=int, default=4)
    parser.add_argument('--gram_prompt_dim', type=int, default=1024)
    parser.add_argument('--style_prompts', type=int, default=4)
    parser.add_argument('--style_prompt_dim', type=int, default=1024)

    args = parser.parse_args()
    return args


root_path = 'DSR/'

image_mean = (0.48145466, 0.4578275, 0.40821073)
image_std = (0.26861954, 0.26130258, 0.27577711)

args = parse_args()

filelist = json.load(open('DSR/train.json', 'r'))

process = image_transform(224, False, image_mean, image_std)

embed = []


if args.prompt == 'ShallowPrompt':
    model = ShallowStyleRetrieval(args)
elif args.prompt == 'DeepPrompt':
    model = DeepStyleRetrieval(args)
elif args.prompt == 'BLIP_Retrieval':
    model = BLIP_Retrieval(args)
else:
    model = BLIP_ITM(args)

model.load_state_dict(torch.load(args.resume,map_location="cpu"))
model = model.to(args.device)

all_center = []
save_path = f"gram_cluster_{args.prompt}" if args.prompt != "ShallowPrompt" else f"gram_cluster_{args.prompt}_{args.prompt_location}"
os.makedirs(save_path,exist_ok=True)
for subdir in ['sketch','art','images','mosaic']:
    for file in tqdm(filelist):

        path = root_path + f'{subdir}/' + file['image']
        img = process(Image.open(path)).unsqueeze(0).to('cuda:0')
        feat = model._get_gram_prompt(img)  # (1, 4, 1024)
        embed.append(feat.detach().cpu())


    feature = torch.stack(embed, dim=0).squeeze(1).detach().numpy()
    feature = feature.reshape(feature.shape[0],-1)

    np.save(f'{save_path}/{subdir}.npy', feature)

    kmeans = KMeans(n_clusters=1)

    kmeans.fit(feature)

    cluster_labels = kmeans.labels_

    cluster_centers = kmeans.cluster_centers_

    cluster_tensor = np.mean(cluster_centers, axis=0)
    cluster_tensor = cluster_tensor.reshape((1, 4096))
    all_center.append(cluster_tensor)

all_save_path = f'style_cluster_{args.prompt}' if args.prompt != "ShallowPrompt" else f'style_cluster_{args.prompt}_{args.prompt_location}'
np.save(f'{all_save_path}.npy', np.concatenate(all_center))
```

**Comment by Hyq0203** [comment]
Thank you, brother. Good luck to you！
